### PR TITLE
api(nvim_open_win): add "noautocmd" option

### LINKED ISSUE #14524: PR #13664 cause floating window close
Is PR #13664 https://github.com/neovim/neovim/commit/5165513198dbbe7415f1b8c51225ae8f822628ec  changed floating widnow API ?

The changes between 4be0e92db0 and 5165513198db broken my plugin https://github.com/ray-x/guihua.lua/

The test: https://github.com/ray-x/guihua.lua/blob/master/lua/guihua/tests/listview_test.lua failed. The test is supposed to create a list view window and with cursor moving, it should create a floating window of preview.  But currently, if the cursor moved, the listview will be destroyed.

@ivechan

### PR DESCRIPTION
**Pull request by seandewar**
Had some spare time, so I thought I'd follow up with an initial implementation of https://github.com/neovim/neovim/pull/13664#issuecomment-838185499 to guage some feedback.

Closes #14524, though plugin authors will need to update things.

### CODE REVIEW

File: src/nvim/api/private/helpers.c [hunk 1]
```diff
@@ -2015,6 +2015,16 @@ bool parse_float_config(Dictionary config, FloatConfig *fconfig, bool reconf,
         api_set_error(err, kErrorTypeValidation,
                       "Invalid value of 'style' key");
       }
+    } else if (strequal(key, "noautocmd") && new_win) {
+      if (val.type == kObjectTypeInteger) {
```
Comment (bfredl) [inline src/nvim/api/private/helpers.c:2019]: can use `api_object_to_bool`

File: src/nvim/window.c [hunk 2]
```diff
@@ -570,6 +570,44 @@ static void cmd_with_count(char *cmd, char_u *bufp, size_t bufsize,
...
+  win_T *win = find_window_by_handle(window, err), *save_curwin = curwin;
+  buf_T *buf = find_buffer_by_handle(buffer, err);
+  tabpage_T *tab = win_find_tabpage(win), *save_curtab = curtab;
+
+  if (!win || !buf) {
+    return;
+  }
+
+  int result
+      = noautocmd
```
Comment (bfredl) [inline src/nvim/window.c:584]: could use `switch_win_noblock` always and instead `if (noautocmd) { block_autocmds(); }` and same on the other side.

File: src/nvim/api/private/helpers.c (hunk 1 above)
Comment (seandewar) [reply src/nvim/api/private/helpers.c]: Would you like me to update the other Boolean keys to use that in another commit?

Comment (bfredl) [reply src/nvim/api/private/helpers.c]: yes if you want to :)

File: src/nvim/api/private/helpers.c [hunk 3]
```diff
@@ -1968,26 +1968,11 @@ bool parse_float_config(Dictionary config, FloatConfig *fconfig, bool reconf,
...
-        fconfig->focusable = val.data.integer;
-      } else if (val.type == kObjectTypeBoolean) {
-        fconfig->focusable = val.data.boolean;
-      } else {
-        api_set_error(err, kErrorTypeValidation,
-                      "'focusable' key must be Boolean");
-        return false;
-      }
+      fconfig->focusable
+          = api_object_to_bool(val, "'focusable' key", false, err);
```
Comment (seandewar) [inline src/nvim/api/private/helpers.c:1975]: As this defaults to `true`, should I pass `true` as the 3rd arg here?

Comment (bfredl) [reply src/nvim/api/private/helpers.c:1975]: yes.

**Comment by seandewar** [comment]
I've rebased in case this is still wanted after #14531.

If the original behavior of `nvim_open_win`is preferred as to not break plugins, I could instead turn this into a (yes)`autocmd` option that unblocks autocommands instead?

**Comment by bfredl** [comment]
@seandewar no the present state of this PR is correct. Floating windows should behave like ordinary windows by default, if a plugin expects a different behaviour it should explicitly opt-in to that.

**Comment by blueyed** [comment]
Note that this needed a change with jedi-vim's (WIP) call-signatures branch, where a floating window is used for displaying call signatures: https://github.com/davidhalter/jedi-vim/pull/652/commits/e38372ffa2ff64713f62ae233c4be01cf13cca43.

The error/issue showed itself however only when using https://github.com/numirias/semshi, since it was confused about the additional BufLeave autocommand that was triggered due to jedi-vim showing the call signatures.

I think this is the expected behavior in general, and only wanted to leave this for reference.

**Comment by tomtomjhj** [comment]
@seandewar  Is it intended that `noautocmd=true` doesn't block `WinNew`? I was expecting it to behave like `:noautocmd split ...`, which blocks every autocmd, including `WinNew`.

The current version for quick reference:
https://github.com/neovim/neovim/blob/15a2dd9e963784123273ec830e82e24ecea4ad0b/src/nvim/api/win_config.c#L294-L303

**Comment by seandewar** [comment]
Yup, it's supposed to block only autocommands fired from setting the current buffer for the window (which was the behaviour before this PR).

**Comment by tomtomjhj** [comment]
I think not blocking `WinNew` makes it difficult to use `WinNew`, and probably renders it useless.

For example, I want to highlight some stuff on window with `matchadd` (which is window-local) based on the state stored in a buffer variable. When `:split`-ing (without argument), I would like to duplicate the highlights for the new window. A natural choice would be to setup a `WinNew` autocmd that reads the buffer variable and calls `matchadd`.

But the problem is that since `WinNew` is fired before setting the buffer for the window, the `WinNew` autocmd cannot know what buffer the new window is meant for. In other words, `WinNew` can't distinguish `:split` and `:split another_buffer`. One solution would be to do `matchadd` in `WinNew` and clear it in the `BufEnter` event when the window displays another buffer. This works in most cases, but it won't work if the autocmd has some non-undoable side effect. But even worse, this suboptimal solution is not even possible for `nvim_open_win(another_buffer, false, {noautocmd=true})`, because it blocks the undoing `BufEnter`. I can't control `nvim_open_win`, because it is invoked by another plugin (nvim-cmp's custom popup menu, which is reasonable use case of `noautocmd=true`).

An alternative solution would be to use `WinEnter` instead of of `WinNew`. But this is suboptimal as it runs whenever the focus moves to another window, which is why (I believe) `WinNew` exists.

So I believe it would be better to block `WinNew` unless there is a reasonable use case for blocking `Buf*` but not `WinEnter`.

**Comment by seandewar** [comment]
Ah, thanks for letting me know about that use-case.

My concern is that blocking `WinNew` also implies that we _might_ want to block `WinEnter` too when `enter = true` (and `noautocmd = true`), as that event also fires before the buffer is set, which may also cause issues for autocommands like your `WinNew` one that is similarly sensitive to which buffer is being shown in the window when `noautocmd = true`; however, #13664 did not affect the behaviour of `WinEnter`, which did trigger before that PR, so that behaviour ought not to change for `noautocmd`.

I think the general problem here is that there's a mismatch between what's expected from `noautocmd`; the original intention here was a way to opt-in to the pre-#13664 behaviour, as that change broke a hefty amount of plugins at the time (looking back at it, maybe that was a mistake :see_no_evil:). If plugins like cmp prefer to instead open a "temporary" window with minimal side-effects like a popup, something like #19832 would be better suited instead (that can be done today via saving/restoring `&eventignore`, or via `:noautocmd` if used from Vim script, but it's not very convenient).

That said, there may be an argument for blocking `WinNew` anyway given that it didn't fire at all before #25096 (which may imply blocking it is more compatible with the pre-#13664 behaviour); I don't have much of an opinion there. (Any thoughts @bfredl?)

**Comment by bfredl** [comment]
A lot of text here. The intention of `noautocmd` is indeed to block autocmds. The documentation might only mention some specific autocmds as only some autocmds was fired when the documentation was written.

I do not know what about `WinNew` that makes is special so it should not be blocked while the others are. (I mean fundamentally, as a designed API, not as discourse about "what plugins are we currently breaking by accident")

**Comment by seandewar** [comment]
Thanks. I'm OK with opening a PR making `noautocmd` 100% consistent and blocking _all_ events from `nvim_open_win`, which would include `WinEnter` which wasn't blocked before (even if some existing scripts in the wild misbehave as a result of that :innocent:).

**Comment by bfredl** [comment]
you could easily enter as a follow-up API call if you want that to happen outside of the `noautocmd`.

**Comment by seandewar** [comment]
Yup, and that has the added benefit of the target buffer being set at that point.

**Pull request merged**